sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/ShuffleExchangeExec.scala

```diff
@@ -83,10 +83,23 @@ trait ShuffleExchangeLike extends Exchange {
 
   protected def mapOutputStatisticsFuture: Future[MapOutputStatistics]
 
+  /**
+   * Cancels the shuffle job with a reason.
+   */
+  final def cancelShuffleJob(reason: String): Unit = {
```

Comment:
I'm not a fan of the copy-paste code duplication between the "with reason" and "without reason" overloads and think that we should instead refactor the common code to a `private final def doCancelShuffleJob(reason: Option(reason))`. Within that method, we can do something like

```scala
reason match {
  case Some(reason) => action.cancel(reason()
  case None => action.cancel()
}
```

or similar.